text/0000-less-pushy-navigate.md

```diff
@@ -0,0 +1,113 @@
+
+```
+// Example 1
+// This would be preferred
+class DetailsScreen extends React.Component {
+  render() {
+    const { navigation } = this.props;
+    return (
+      <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
+        <Text>Details Screen</Text>
+        <Button
+          title="Go to Details... again"
+          onPress={() => {
+            navigation.push('Details');
+          }}
+        />
+      </View>
+    );
+  }
+}
+```
+
+```
+// Example 2
+// This is contrived, it would make no sense to push a new key like this
+class DetailsScreen extends React.Component {
+  render() {
+    const { navigation } = this.props;
+    return (
+      <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
+        <Text>Details Screen</Text>
+        <Button
+          title="Go to Details... again"
+          onPress={() => {
+            navigation.navigate({
+              routeName: 'Details',
+              key: navigation.getParam('n', 0) + 1
+            });
+          }}
+        />
+      </View>
+    );
+  }
+}
+```
+
+
+# Motivation
+
+- Getting the key for the initial route in a stack is causing people problems (https://github.com/react-navigation/react-navigation/pull/3540). It's easy to push duplicate screens because of this (https://github.com/react-navigation/react-navigation/issues/3265).
+- This also helps mitigate the longstanding issue of double-pushing routes in a `StackNavigator`. This is because you need to explicitly provide a key if you would like to push a route that you expect can have multiple instances in the stack. Many routes aren't like this, for example in the iOS settings app it only makes sense to have one instance of any of the settings screens. Routes that can be pushed multiple times are almost always provided with some params that are used to populate the data on the screen, and some id from that data can be used as the route key (for example, you might navigate to a profile screen with a `userId` param, and you can use that as the key). (https://github.com/react-navigation/rfcs/issues/16)
+- I suspect that this new behavior is more intuitive. `navigate` is more clearly distinguished from `push` -- it tries to go to an existing route and only pushes one if it's not already there. `push` always pushes a new route to state. It makes `navigate` for `StackRouter` behave more similarly to `TabRouter` also.
+
+# Detailed design
+
+We need to change [this branch](https://github.com/react-navigation/react-navigation/blob/2bb91a6740e2f1ade012906dc03c972ba6375cc3/src/routers/StackRouter.js#L185-L190) of StackRouter. The logic will be similar to the [`if (action.key)`](https://github.com/react-navigation/react-navigation/blob/2bb91a6740e2f1ade012906dc03c972ba6375cc3/src/routers/StackRouter.js#L200) branch, but rather than simply match for key we'll match for `routeName` within `state.routes`.
```

Comment:
One question/concern I have: if you're navigating back to an existing routeName, are the other routes expected to be popped? Because it is a stack, I would expect so, but it is possible to change the index and leave the other routes remaining.

Say we have a stack of `A B C*` where the `*` indicates focus index. When we navigate to a routeName of `B`, do we:

- `A B*` - Fully pop the remaining route(s)?
- `A B* C` - Leave the routes but change the active index? (This is the behavior of tab router)
- `A B C B*` - Consider "navigate" to always go forward, which preserves the stack behavior

Reply:
> One question/concern I have: if you're navigating back to an existing routeName, are the other routes expected to be popped? Because it is a stack, I would expect so, but it is possible to change the index and leave the other routes remaining

yeah, we would pop the remaining routes! 

`A B C*` -> navigate to `B` -> `A B*`

custom routers would be expected to implement handling for the navigation with this same spirit however that applies to their situation. it could be thought of as "jump to this route" which would be interpreted by tab as just switching index, and by stack as either pushing it if it's not in state or popping routes in state until it's focused.